### PR TITLE
Game loop now uses display.update_frequency to use variable/fixed frame rate

### LINKED ISSUE #3210: Fixed vs. non-fixed engine timestep (DEF-3802)
With variable refresh rate monitors, gaming devices, phone battery save mode and all web browsers, it's becoming increasingly difficult to determine what correct fixed timestep to use in the engine.

We should look at using actual measured frame time instead of a fixed time step.

Design:(LINK REMOVED)

### LINKED ISSUE #3441: Variable DT does not work on high frame rates (DEF-3337)
Description:
There are a multitude of problems related to vertical sync on PC.
This specific lira only relates to how we calculate frame time.

In DEF-3146 we reverted from QueryPerformanceCounter since that returns irregular results and is not reliable across threads, but in doing so the resolution of the timer dropped so bad that it does not calculate "dt" when the frame rate is high (vsync is off).

We probably need to use QueryPerformaceCounter specifically in the main loop to get enough precision but if we are to do that we need to pin the thread to a specific thread.

(LINK REMOVED)

### LINKED ISSUE #6094: Variable DT As Independent Option
**Is your feature request related to a problem? Please describe (REQUIRED):**
Currently dt being constant or variable depends on whether vsync is on or off, as well as if your game's framerate is uncapped. This means that you cannot have variable dt in a game that uses vsync or a capped framerate, which will result in your game physically slowing down if its framerate is low or experiences lag.

**Describe the solution you'd like (REQUIRED):**
Make variable dt a separate, independent option from vsync and frame cap. With this option checked, dt will be based on time since the last frame (variable) regardless of what vsync and frame cap are set to. That way you can have a game that both uses vsync (to avoid tearing issues) or capped framerate, and maintains its speed when under stress by increasing delta time appropriately.

**Describe alternatives you've considered (REQUIRED):**
Another possibility is introducing two separate update loops, one variable and one fixed. Of course this would require major revisions to the core of the engine.

### LINKED ISSUE #6107: dt = 1 if GPU VSync is turned off and FPS is ultra high
**Describe the bug + How to Reproduce**
I like to develop games with Defold with the following settings:
image

The combination enables "variable dt" mode.

I found that the engine **sets the "dt" variable equals to 1** if:
1. A user turned off VSync for all applications, i.e.
image
2. Game's FPS is more than 1000.

The result is that everything animated/moved according to the dt=1, i.e. animations end almost instantly. Take a look at the video _(green-coloured text is FPS)_:

https://user-images.githubusercontent.com/193258/137761885-13af3a89-eca0-487b-9ea3-513c4376d4e2.mp4

**Expected behavior (REQUIRED)**
The game should be able to run well under any conditions and user system/GPU settings.

**Defold version (REQUIRED):**
  - Version: 1.2.188

**Platforms (REQUIRED):**
 - Platforms: Windows
 - GPU: Radeon 6600 XT

**Minimal repro case project (OPTIONAL):**
[issue_vsync_off_high_dt.zip](https://github.com/defold/defold/files/7366398/issue_vsync_off_high_dt.zip)

**Logs (OPTIONAL):**
image

### PR DESCRIPTION
**Pull request by JCash**
The display.vsync is now deprecated, in favor of display.swap_interval

`display.swap_interval` (integer) Sets the OpenGL swap interval (new property)

`display.vsync` - deprecated. Set to false to force a `display.swap_interval = 0`

`display.update_frequency`
 - 0 means variable frame rate
 - != 0 means fixed frame rate. Capped at runtime towards the actual frame rate. I.e. you cannot update the game twice in a frame.

Message `set_update_frequency` to control the "display.update_frequency" at runtime.

### CODE REVIEW

File: engine/engine/src/engine.cpp [hunk 1]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
         return memcount;
     }
 
-    void Step(HEngine engine)
+    static void StepFrame(HEngine engine, float dt)
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1428]: The code here is the old contents of Engine::Step()
As evident, a lot of the stuff isn't really something we'd like to run more than once per engine frame.
(e.e HID update, Flip() etc) but for now this is what we have.

File: engine/engine/src/engine.cpp [hunk 2]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
+        }
+        dmProfile::Release(profile);
 
-                dmGraphics::Flip(engine->m_GraphicsContext);
+        ++engine->m_Stats.m_FrameCount;
+    }
 
-                engine->m_FlipTime = dmTime::GetTime();
-                engine->m_PreviousRenderTime = engine->m_FlipTime - flip_time_start;
+    static void CalcTimeStep(HEngine engine, float& step_dt, uint32_t& num_steps)
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1656]: A separate function to make it clearer how we produce the delta times.

File: engine/engine/src/engine.cpp [hunk 3]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
-                        uint32_t width = dmGraphics::GetWidth(engine->m_GraphicsContext);
-                        uint32_t height = dmGraphics::GetHeight(engine->m_GraphicsContext);
-                        uint32_t buffer_size = width * height * 4;
+        float frame_dt = (float)(frame_time / 1000000.0);
 
-                        dmGraphics::ReadPixels(engine->m_GraphicsContext, record_data->m_Buffer, buffer_size);
+        // Never allow for large hitches
+        if (frame_dt > 0.5f) {
+            frame_dt = 0.5f;
+        }
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1667]: Good to know about this cap.
We might want is as a setting, but until anyone requests it...
SHould probably be documented though.

File: engine/engine/src/engine.cpp [hunk 4]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
-                    record_data->m_FrameCount++;
-                }
-            }
-            dmProfile::Release(profile);
+        // Variable frame rate
+        if (engine->m_UpdateFrequency == 0)
+        {
+            step_dt = frame_dt;
+            num_steps = 1;
+            return;
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1674]: The default behavior: run as fast as we can given the OS vsync settings, and also the swap interval

File: engine/engine/src/engine.cpp [hunk 5]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
+
+        // We don't allow having a higher framerate than the actual variable frame rate
+        // since the update+render is currently coupled together and also Flip() would be called more than once.
+        if (fixed_dt < frame_dt)
+        {
+            step_dt = frame_dt;
+            num_steps = 1;
+            engine->m_AccumFrameTime = 0; // we consumed it all
+            return;
+        }
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1688]: If we tried to run faster than the engine frame rate, we cap to the engine frame rate.

File: engine/engine/src/engine.cpp [hunk 6]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
+
+        engine->m_AccumFrameTime += frame_dt;
 
+        float num_steps_f = engine->m_AccumFrameTime / fixed_dt;
 
-            ++engine->m_Stats.m_FrameCount;
+        num_steps = (uint32_t)num_steps_f;
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1694]: Given the early outs as seen above, the number of steps can only be 0 or one here.

File: engine/engine/src/engine.cpp [hunk 7]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
+        CalcTimeStep(engine, step_dt, num_steps);
+
+        for (uint32_t i = 0; i < num_steps; ++i)
+        {
+            // We currently cannot separate the update from the render,
+            // since some of the update is done in the render updates (e.g. sprite transforms)
+            StepFrame(engine, step_dt);
+
+            if (!engine->m_Alive)
+                break;
```
Comment (JCash) [inline engine/engine/src/engine.cpp:1718]: The simplifies Step() function.

File: engine/engine/src/engine.cpp [hunk 8]
```diff
@@ -852,48 +844,24 @@ namespace dmEngine
...
-        uint32_t setting_update_frequency = dmConfigFile::GetInt(engine->m_Config, "display.update_frequency", 0);
-        uint32_t update_frequency = setting_update_frequency;
-        uint32_t swap_interval = 1;
+        engine->m_UpdateFrequency = dmConfigFile::GetInt(engine->m_Config, "display.update_frequency", 0);
+
+        SetUpdateFrequency(engine, engine->m_UpdateFrequency);
+
+        bool setting_vsync = dmConfigFile::GetInt(engine->m_Config, "display.vsync", true); // Deprecated
 
+        uint32_t opengl_swap_interval = dmConfigFile::GetInt(engine->m_Config, "display.swap_interval", 1);
```
Comment (britzl) [inline engine/engine/src/engine.cpp:859]: Is this only for opengl? What about vulkan?

File: engine/engine/src/engine.cpp [hunk 9]
```diff
@@ -1466,294 +1425,297 @@ namespace dmEngine
...
-                            // might have entered background at this point and OpenGL calls are not permitted and will
-                            // crash the application
-                            dmProfile::Release(profile);
-                            return;
-                        }
-                    }
-                    /* Script context updates */
+                    // Call pre render functions for extensions, if available.
+                    // We do it here before we render rest of the frame
+                    // if any extension wants to render on under of the game.
```
Comment (britzl) [inline engine/engine/src/engine.cpp:1544]: "render on under of the game" sounds a bit strange. Perhaps "render below" or "render before"

**Review by britzl — DISMISSED** [review]
Looks good. Only a question about opengl swap interval .

File: engine/engine/src/engine.cpp (hunk 8 above)
Comment (JCash) [reply engine/engine/src/engine.cpp:859]: You cannot set swap interval on Vulkan.
There are different present modes for vulkan, and we currently use the FIFO mode. We might want to expose that as a setting later on (when someone requests it)

Comment (JCash) [reply engine/engine/src/engine.cpp:859]: I feel there's a danger in coupling the two backend behaviors too closely.

Comment (britzl) [reply engine/engine/src/engine.cpp:859]: Ok, got it. Makes sense!

**Review by britzl — DISMISSED** [review]

**Pull request merged**